File: impls/python/step2_eval.py
from reader import read_str
from printer import pr_str


def add(val_a, val_b):
    return {
            "type": "number",
            "value": val_a["value"] + val_b["value"]
    }

def sub(val_a, val_b):
    return {
            "type": "number",
            "value": val_a["value"] - val_b["value"]
    }

def mult(val_a, val_b):
    return {
            "type": "number",
            "value": val_a["value"] * val_b["value"]
    }

def div(val_a, val_b):
    return {
            "type": "number",
            "value": val_a["value"] / val_b["value"]
    }

repl_env = {
        '+': {"type": "fn", "value": add},
        '-': {"type": "fn", "value": sub},
        '*': {"type": "fn", "value": mult},
        '/': {"type": "fn", "value": div},
}

def eval_ast(ast, env):
    if ast["type"] == "symbol":
        try:
            return env[ast["value"]]
        except:
            return {
                "type": "error",
                "value": {
                    "type": "string",
                    "value": "'" + ast["value"] + "' not found"
                    }
            }
    if ast["type"] in ("list", "vector"):
        items = map(lambda ast: eval(ast, env), ast["value"])
        return {
            "type": ast["type"],
            "value": items,
            }
    if ast["type"] == "hashmap":
        new_val = {}
        for key, value in ast["value"].items():
            new_val[key] = eval(value, env)
        return {
            "type": "hashmap",
            "value": new_val
            }
    return ast

def eval(ast, env):
    if ast["type"] != "list":
        return eval_ast(ast, env)
    if len(ast["value"]) == 0:
        return ast
    new_ast = eval_ast(ast, env)
    fn = new_ast["value"][0]

    if fn["type"] == "error":
        return fn

    return fn["value"](*new_ast["value"][1:])

def rep(line):
    print(pr_str(eval(read_str(line), repl_env)))

while True:
    line = raw_input("user> ")
    if (len(line) != 0):
        rep(line)

File: impls/rust/step2_eval.rs
mod reader;
use reader::MalVal;
use std::collections::HashMap;
use rustyline::Editor;


fn eval_ast(ast: MalVal, env: HashMap<&str, MalVal>) -> MalVal {
    match ast {
        MalVal::MalSymbol(symbol) => {
            env.get(symbol).or_else(MalVal::MalError(Box::new(MalVal::MalString("Not found".to_string())))),
        },
        MalVal::MalList(content) => {
            let items = content.as_ref.to_vec.map(|item| => EVAL(item, env));
            MalVal::MalList(Box::new(items))
        },
        val => val,
    }
}

fn EVAL(ast: MalVal, env: HashMap<&str, MalVal>) -> MalVal {
    match ast {
        MalVal::MalList(content) => {
            if (&content.len() == 0) {
                ast
            } else {
                match eval_ast(ast, env) {
                    MalVal::MalList(content) => {
                        let items = content.as_ref.to_vec.map(|item| => EVAL(item, env));
                        MalVal::MalList(Box::new(items))
                    },
                    _ => MalVal::MalError(Box::new(MalVal::MalString("unexpected".to_string()))),
                }
            }
        },
        _ => eval_ast(ast, env),

    }
    return ast;
}

fn rep(line: &str, env: HashMap<&str, MalVal>) {
    let val = reader::read_str(&line);
    let evaluated = EVAL(val, env);
    println!("{}", evaluated.print(true));
}

fn main() {
    let mut rl = Editor::<()>::new();
    loop {
        let readline = rl.readline("user> ");
        let mut repl_env = HashMap::new();

        repl_env.insert("+",
            MalVal::MalFunc(|args: &[MalVal]| {
                match (&args[0], &args[1]) {
                    (MalVal::MalNumber(a), MalVal::MalNumber(b)) => MalVal::MalNumber(a + b),
                    _ => MalVal::MalError(Box::new(MalVal::MalString("Wrong operation".to_string())))
                }
            }));

        match readline {
            Ok(line) => {
                rep(&line, repl_env);
            },
            Err(_) => {
                break
            },
        }
    }
}
